### PR TITLE
IDEMPIERE-4567 Resetting payment allocation to a charge leaves wrong …

### PR DESCRIPTION
**Pull request by CarlosRuiz-globalqss**
…BP Balance

MAllocationHdr.updateBP was broken, the form Payment Allocation worked just because it was calling bpartner.setTotalOpenBalance() at the end
so, I refactored the MAllocationHdr.updateBP and before/afterDelete to call bpartner.setTotalOpenBalance instead if the broken algorithm

Unit test added for allocating and deleting a customer invoice, a vendor invoice, and a charge

https://idempiere.atlassian.net/browse/IDEMPIERE-4567

### CODE REVIEW

File: org.adempiere.base/src/org/compiere/model/MAllocationHdr.java [hunk 1]
```diff
@@ -306,6 +304,8 @@ protected boolean beforeSave (boolean newRecord)
 		return true;
 	}	//	beforeSave
 
+	private List<Integer> m_bps_beforeDelete = Collections.synchronizedList(new ArrayList<Integer>());;
```
Comment (hengsin) [inline org.adempiere.base/src/org/compiere/model/MAllocationHdr.java:307]: double semicolon here.

File: org.adempiere.base/src/org/compiere/model/MAllocationHdr.java [hunk 2]
```diff
@@ -327,17 +327,37 @@ protected boolean beforeDelete ()
 
 		//	Unlink
 		getLines(true);
-		if (!updateBP(true)) 
+		String prevStatus = getDocStatus();
+		setDocStatus(STATUS_Reversed);
+		if (!updateBP())
 			return false;
+		setDocStatus(prevStatus);
```
Comment (hengsin) [inline org.adempiere.base/src/org/compiere/model/MAllocationHdr.java:334]: this one is hard to understand - why this docstatus stuff is needed in beforeDelete ?

File: org.adempiere.base/src/org/compiere/model/MAllocationHdr.java [hunk 3]
```diff
@@ -327,17 +327,37 @@ protected boolean beforeDelete ()
 
 		//	Unlink
 		getLines(true);
-		if (!updateBP(true)) 
+		String prevStatus = getDocStatus();
+		setDocStatus(STATUS_Reversed);
+		if (!updateBP())
```
Comment (hengsin) [inline org.adempiere.base/src/org/compiere/model/MAllocationHdr.java:332]: why the need to call updateBP() here if setTotalOpenBalance is call for each bp in afterDelete() ?

**Review by hengsin — CHANGES_REQUESTED** [review]
added some questions.

**Comment by CarlosRuiz-globalqss** [comment]
Thanks @hengsin,

The extra code in beforeDelete was leftovers from a previous try that didn't work.  Cleared.

Regards,

Carlos Ruiz

**Review by hengsin — APPROVED** [review]
looks good.

**Pull request merged**